pkg/store/multitsdb.go

```diff
@@ -0,0 +1,284 @@
+	logger     log.Logger
+	component  component.SourceStoreAPI
+	tsdbStores func() map[string]*TSDBStore
+}
+
+// NewMultiTSDBStore creates a new TSDBStore.
+func NewMultiTSDBStore(logger log.Logger, _ prometheus.Registerer, component component.SourceStoreAPI, tsdbStores func() map[string]*TSDBStore) *MultiTSDBStore {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	return &MultiTSDBStore{
+		logger:     logger,
+		component:  component,
+		tsdbStores: tsdbStores,
+	}
+}
+
+// Info returns store information about the Prometheus instance.
+func (s *MultiTSDBStore) Info(ctx context.Context, req *storepb.InfoRequest) (*storepb.InfoResponse, error) {
+	stores := s.tsdbStores()
+
+	resp := &storepb.InfoResponse{
+		StoreType: s.component.ToProto(),
+	}
+	if len(stores) == 0 {
+		return resp, nil
+	}
+
+	infos := make([]*storepb.InfoResponse, 0, len(stores))
+	for _, store := range stores {
+		info, err := store.Info(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		infos = append(infos, info)
+	}
+
+	resp.MinTime = infos[0].MinTime
+	resp.MaxTime = infos[0].MaxTime
+
+	for i := 1; i < len(infos); i++ {
+		if resp.MinTime > infos[i].MinTime {
+			resp.MinTime = infos[i].MinTime
+		}
+		if resp.MaxTime < infos[i].MaxTime {
+			resp.MaxTime = infos[i].MaxTime
+		}
+	}
+
+	// We can rely on every underlying TSDB to only have one labelset, so this
+	// will always allocate the correct length immediately.
+	resp.LabelSets = make([]storepb.LabelSet, 0, len(infos))
+	for _, info := range infos {
+		resp.LabelSets = append(resp.LabelSets, info.LabelSets...)
+	}
+
+	return resp, nil
+}
+
+type seriesSetServer struct {
+	grpc.ServerStream
+
+	ctx context.Context
+
+	warnCh warnSender
+	recv   chan *storepb.Series
+	cur    *storepb.Series
+
+	errMtx *sync.Mutex
+	err    error
+}
+
+func newSeriesSetServer(
+	ctx context.Context,
+	warnCh warnSender,
+) *seriesSetServer {
+	return &seriesSetServer{
+		ctx:    ctx,
+		warnCh: warnCh,
+		recv:   make(chan *storepb.Series),
+		errMtx: &sync.Mutex{},
+	}
+}
+
+func (s *seriesSetServer) Context() context.Context {
+	return s.ctx
+}
+
+func (s *seriesSetServer) Run(store *TSDBStore, r *storepb.SeriesRequest) {
+	err := store.Series(r, s)
+	if err != nil {
+		if r.PartialResponseDisabled {
+			s.errMtx.Lock()
+			s.err = err
+			s.errMtx.Unlock()
+		} else {
+			s.warnCh.send(storepb.NewWarnSeriesResponse(err))
+		}
+	}
+
+	close(s.recv)
+}
+
+func (s *seriesSetServer) Send(r *storepb.SeriesResponse) error {
+	series := r.GetSeries()
+	chunks := make([]storepb.AggrChunk, len(series.Chunks))
+	copy(chunks, series.Chunks)
+	s.recv <- &storepb.Series{
+		Labels: series.Labels,
+		Chunks: chunks,
+	}
+	return nil
+}
+
+func (s *seriesSetServer) Next() (ok bool) {
+	s.cur, ok = <-s.recv
+	return ok
+}
+
+func (s *seriesSetServer) At() ([]storepb.Label, []storepb.AggrChunk) {
+	if s.cur == nil {
+		return nil, nil
+	}
+	return s.cur.Labels, s.cur.Chunks
+}
+
+func (s *seriesSetServer) Err() error {
```

Comment:
just curious why only Err have mtxs?

Reply:
It doesn't. It's a leftover from intermediate work. Good catch.